### PR TITLE
Add composer_sort as a tag

### PR DESCRIPTION
**Pull request by dosoe**
Already discussed on https://github.com/beetbox/beets/issues/2519
This is my first pull request (almost) and my first contribution to any open-source project, so please be kind... 
It's heavily using https://github.com/beetbox/beets/pull/2333/files/7060b512b848fb36f49dcc091f65a6e7baa83d4f and https://picard.musicbrainz.org/docs/mappings/ . I would have liked to add an arranger_sort tag, but arranger_sort is not listed in the tag mapping (unlike composer_sort), so it will be more difficult.

### CODE REVIEW

**Comment by dosoe** [comment]
It works fine on my files, but I didn't do any further testing.

**Comment by dosoe** [comment]
Concerning the ID3v2 tag, it is TSOC (Picard>=1.3) and TXXX:COMPOSERSORT (Picard<=1.2). Since the official picard version is now 1.4.2, I used only the second, but that might cause problems. Concerned are  beets/mediafile.py

File: beets/autotag/__init__.py [hunk 1]
```diff
@@ -49,8 +49,12 @@ def apply_item_metadata(item, track_info):
         item.lyricist = track_info.lyricist
     if track_info.composer is not None:
         item.composer = track_info.composer
+    if track_info.composer_sort is not None:
+        item.composer_sort = track_info.composer_sort
     if track_info.arranger is not None:
         item.arranger = track_info.arranger
+    
+
```
Comment (Freso) [inline beets/autotag/__init__.py]: Just remove these blank lines.

File: beets/autotag/hooks.py [hunk 2]
```diff
@@ -155,8 +156,8 @@ def __init__(self, title, track_id, artist=None, artist_id=None,
                  length=None, index=None, medium=None, medium_index=None,
                  medium_total=None, artist_sort=None, disctitle=None,
                  artist_credit=None, data_source=None, data_url=None,
-                 media=None, lyricist=None, composer=None, arranger=None,
-                 track_alt=None):
+                 media=None, lyricist=None, composer=None, composer_sort=None,
```
Comment (Freso) [inline beets/autotag/hooks.py]: There's a trailing space here.

File: beets/mediafile.py [hunk 3]
```diff
@@ -1632,12 +1632,21 @@ def update(self, dict):
         StorageStyle('LYRICIST'),
         ASFStorageStyle('WM/Writer'),
     )
+
```
Comment (Freso) [inline beets/mediafile.py]: Follow the style of the rest of the code and don't add a blank line here.

File: beets/mediafile.py [hunk 4]
```diff
@@ -1632,12 +1632,21 @@ def update(self, dict):
...
         ASFStorageStyle('WM/Writer'),
     )
+    
     composer = MediaField(
         MP3StorageStyle('TCOM'),
         MP4StorageStyle('\xa9wrt'),
         StorageStyle('COMPOSER'),
         ASFStorageStyle('WM/Composer'),
     )
+
```
Comment (Freso) [inline beets/mediafile.py]: Or here.

File: beets/mediafile.py [hunk 5]
```diff
@@ -1632,12 +1632,21 @@ def update(self, dict):
...
         ASFStorageStyle('WM/Composer'),
     )
+    
+    composer_sort = MediaField(
+        MP3StorageStyle('TSOC'),
+        MP4StorageStyle('soco'),
+        StorageStyle('COMPOSERSORT'),
+        ASFStorageStyle('WM/Composersortorder'),
+    )
+
```
Comment (Freso) [inline beets/mediafile.py]: Or here.

File: test/_common.py [hunk 6]
```diff
@@ -68,6 +68,7 @@ def item(lib=None):
         genre=u'the genre',
         lyricist=u'the lyricist',
         composer=u'the composer',
+        composer_sort=u'the sortname of the composer',
```
Comment (Freso) [inline test/_common.py]: There's no `artist_sort` here AFAICT, are you sure `composer_sort` should be here?

File: test/test_mediafile.py [hunk 7]
```diff
@@ -141,6 +141,12 @@ def test_add_image_structure(self):
                       type=ImageType.composer)
         mediafile.images += [image]
         mediafile.save()
+
```
Comment (Freso) [inline test/test_mediafile.py]: Don't have whitespace on (otherwise) empty lines.

File: test/test_mediafile.py [hunk 8]
```diff
@@ -141,6 +141,12 @@ def test_add_image_structure(self):
                       type=ImageType.composer)
         mediafile.images += [image]
         mediafile.save()
+        
+        image = Image(data=self.png_data, desc=u'the sortname of the composer',
+                      type=ImageType.composer_sort)
+        mediafile.images += [image]
+        mediafile.save()
+
```
Comment (Freso) [inline test/test_mediafile.py]: No need for double blank lines here.

File: test/test_mediafile.py [hunk 9]
```diff
@@ -141,6 +141,12 @@ def test_add_image_structure(self):
                       type=ImageType.composer)
         mediafile.images += [image]
         mediafile.save()
+        
+        image = Image(data=self.png_data, desc=u'the sortname of the composer',
+                      type=ImageType.composer_sort)
+        mediafile.images += [image]
+        mediafile.save()
```
Comment (Freso) [inline test/test_mediafile.py]: AFAICT, this is a test relating to images. I'm not sure `composer_sort` stuff is relevant here.

File: test/test_mediafile.py [hunk 10]
```diff
@@ -150,6 +156,12 @@ def test_add_image_structure(self):
         self.assertExtendedImageAttributes(
             image, desc=u'the composer', type=ImageType.composer
         )
+
```
Comment (Freso) [inline test/test_mediafile.py]: Don't include whitespace on blank lines.

File: test/test_mediafile.py [hunk 11]
```diff
@@ -150,6 +156,12 @@ def test_add_image_structure(self):
         self.assertExtendedImageAttributes(
             image, desc=u'the composer', type=ImageType.composer
         )
+        
+        images = (i for i in mediafile.images if i.desc == u'the sortname of the composer')
+        image = next(images, None)
+        self.assertExtendedImageAttributes(
+            image, desc=u'the sortname of the composer', type=ImageType.composer_sort
+        )
```
Comment (Freso) [inline test/test_mediafile.py]: Again, not sure this test is relevant here.

File: test/test_mediafile.py [hunk 12]
```diff
@@ -193,6 +205,11 @@ def test_add_tiff_image(self):
                       type=ImageType.composer)
         mediafile.images += [image]
         mediafile.save()
+
```
Comment (Freso) [inline test/test_mediafile.py]: Whitespace, blank line…

File: test/test_mediafile.py [hunk 13]
```diff
@@ -202,6 +219,11 @@ def test_add_tiff_image(self):
                      mediafile.images))[0]
         self.assertExtendedImageAttributes(
             image, desc=u'the composer', type=ImageType.composer)
+
```
Comment (Freso) [inline test/test_mediafile.py]: Whitespace, blank line…

File: test/test_mediafile.py [hunk 14]
```diff
@@ -318,29 +340,30 @@ class ReadWriteTestBase(ArtTestMixin, GenreListTestMixin,
...
-        'mb_albumid':  '9e873859-8aa4-4790-b985-5a953e8ef628',
-        'mb_artistid': '7cf0ea9d-86b9-4dad-ba9e-2355a64899ea',
-        'art':         None,
-        'label':       u'the label',
+        'title':         u'full',
+        'artist':        u'the artist',
+        'album':         u'the album',
+        'genre':         u'the genre',
+        'composer':      u'the composer',
+        'composer_sort': u'the sortname of the composer',
```
Comment (Freso) [inline test/test_mediafile.py]: I don't see `artist_sort` here, so maybe no need for `composer_sort` either?

Comment (dosoe) [inline test/test_mediafile.py]: true, but I wonder why there isn't artist_sort . Imo where there is a name there also should be the sort name.

**Comment by dosoe** [comment]
I have nothing to do with the tests, they just popped randomly.

File: test/test_mediafile.py (hunk 14 above)
Comment (dosoe) [inline test/test_mediafile.py]: anyway, removed it.

**Comment by dosoe** [comment]
'Also, for good measure, consider making a commit updating the changelog too. :)' What does that mean?

**Comment by dosoe** [comment]
and thanks for the links.

**Comment by dosoe** [comment]
by the way, as a familiar of musicbrainz, you might know how to add a "work" tag properly.

File: test/test_mediafile.py [hunk 15]
```diff
@@ -373,7 +350,6 @@ class ReadWriteTestBase(ArtTestMixin, GenreListTestMixin,
         'genre',
         'lyricist',
         'composer',
-        'composer_sort',
```
Comment (Freso) [inline test/test_mediafile.py]: This list *does* have an `artist_sort`, so the `composer_sort` is probably also applicable here.

**Comment by Freso** [comment]
> I have nothing to do with the tests, they just popped randomly.

That is not how tests work, sorry. The tests break either because changes in this pull request broke something that make existing tests fail, or because new tests added in the pull request are failing. If you really think there's randomness involved, you "Restart build" in Travis (not sure if you can manually trigger a rebuild in AppVeyor). The "Details" links under the "All checks have failed" should have more details you can use to try and figure out what code is failing once you've been convinced it isn't a fluke.

**Comment by dosoe** [comment]
Ok, sorry, I should have been more clear. I know that these tests are not strictly speaking random, I just wanted to say that I have nothing to do with them (and btw I didn't add any test in this request).

File: beets/autotag/mb.py [hunk 16]
```diff
@@ -218,10 +219,12 @@ def track_info(recording, index=None, medium=None, medium_index=None,
                     lyricist.append(artist_relation['artist']['name'])
                 elif type == 'composer':
                     composer.append(artist_relation['artist']['name'])
+                    composer_sort.append(artist_relation['artist']['sort-name'])
```
Comment (jackwilsdon) [inline beets/autotag/mb.py]: This line is now too long, which is why our [linting test](https://travis-ci.org/beetbox/beets/jobs/227173206) is failing. You'll need to wrap it to be under 80 characters, the easiest way being this:

```Python
composer_sort.append(
    artist_relation['artist']['sort-name'])
```

**Comment by sampsyo** [comment]
Thank you! :sparkles: All merged up.

**Comment by dosoe** [comment]
My first merge!

**Pull request merged**